rotkehlchen/chain/evm/decoding/balancer/v3/decoder.py

```diff
@@ -167,11 +178,105 @@ def _order_lp_events(
         )
         return decoded_events
 
+    def _process_swap_events(
+            self,
+            transaction: 'EvmTransaction',
+            decoded_events: list['EvmEvent'],
+            all_logs: list['EvmTxReceiptLog'],
+    ) -> list['EvmEvent']:
+        """Process Balancer v3 swap events by consolidating transfers into proper trade events.
+
+        Reads swap logs to identify which tokens are involved in swaps, then transforms
+        corresponding transfer events into trade events. Removes duplicate events for the same token.
+        """  # noqa: E501
+        in_assets, out_assets = set(), set()
+        for tx_log in all_logs:
+            if tx_log.topics[0] != SWAP_TOPIC:
+                continue
+
+            # track involved assets only, not amounts. wrapping/unwrapping operations
+            # occur before swaps, so using swap log amounts would miss wrapped token portions
+            # and only capture the direct swap amounts, leading to incomplete event matching
+            if (token_out := self.base.get_or_create_evm_asset(bytes_to_address(tx_log.topics[2]))) == CHAIN_TO_WRAPPED_TOKEN[self.evm_inquirer.blockchain]:  # noqa: E501
+                out_assets.add(self.evm_inquirer.native_token)
+            if (token_in := self.base.get_or_create_evm_asset(bytes_to_address(tx_log.topics[3]))) == CHAIN_TO_WRAPPED_TOKEN[self.evm_inquirer.blockchain]:  # noqa: E501
+                in_assets.add(self.evm_inquirer.native_token)
+
+            in_assets.add(token_in)
+            out_assets.add(token_out)
+
+        # collect amounts, find events to transform, and identify duplicates
+        out_amounts: defaultdict[Asset, FVal] = defaultdict(lambda: ZERO)
+        in_amounts: defaultdict[Asset, FVal] = defaultdict(lambda: ZERO)
+        events_to_remove = []
+        processed_tokens, in_event, out_event = set(), None, None
+
+        for event in decoded_events:
+            if (
+                event.event_type == HistoryEventType.SPEND and
+                event.event_subtype == HistoryEventSubType.NONE and
+                event.asset in out_assets
+            ):
+                out_amounts[event.asset] += event.amount
+                if event.asset in processed_tokens:
+                    events_to_remove.append(event)
+                else:
+                    out_event = event
+                    processed_tokens.add(event.asset)
```

Comment:
```suggestion
                if out_event is not None:
                    out_event.amount += event.amount
                    events_to_remove.append(event)
                else:
                    out_event = event
```
Would something like this work? Or will this get an event with the wrong asset in some cases? With the current in/out amounts dicts and the processed_tokens list I guess it ends up using the event from the last token to be encountered rather than the first? Its a bit confusing. This should at least be explained in a comment somewhere here.